controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,388 @@
+		jsonData, err := json.MarshalIndent(gw, "", "  ")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.V(1).Info(string(jsonData))
+	}
+
+	desired, err := r.desiredRateLimitingWASMPlugin(ctx, gw)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcileResource(ctx, &istioclientgoextensionv1alpha1.WasmPlugin{}, desired, rlptools.WASMPluginMutator)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Rate limiting WASMPlugin reconciled successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) desiredRateLimitingWASMPlugin(ctx context.Context, gw *gatewayapiv1.Gateway) (*istioclientgoextensionv1alpha1.WasmPlugin, error) {
+	baseLogger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &istioclientgoextensionv1alpha1.WasmPlugin{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "WasmPlugin",
+			APIVersion: "extensions.istio.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rlptools.WASMPluginName(gw),
+			Namespace: gw.Namespace,
+		},
+		Spec: istioextensionsv1alpha1.WasmPlugin{
+			Selector:     kuadrantistioutils.WorkloadSelectorFromGateway(ctx, r.Client(), gw),
+			Url:          rlptools.WASMFilterImageURL,
+			PluginConfig: nil,
+			// Insert plugin before Istio stats filters and after Istio authorization filters.
+			Phase: istioextensionsv1alpha1.PluginPhase_STATS,
+		},
+	}
+
+	logger := baseLogger.WithValues("wasmplugin", client.ObjectKeyFromObject(wasmPlugin))
+
+	pluginConfig, err := r.wasmPluginConfig(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	if pluginConfig == nil || len(pluginConfig.RateLimitPolicies) == 0 {
+		logger.V(1).Info("pluginConfig is empty. Wasmplugin will be deleted if it exists")
+		utils.TagObjectToDelete(wasmPlugin)
+		return wasmPlugin, nil
+	}
+
+	pluginConfigStruct, err := pluginConfig.ToStruct()
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin.Spec.PluginConfig = pluginConfigStruct
+
+	// controller reference
+	if err := r.SetOwnerReference(gw, wasmPlugin); err != nil {
+		return nil, err
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) wasmPluginConfig(ctx context.Context, gw *gatewayapiv1.Gateway) (*wasm.Plugin, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &wasm.Plugin{
+		FailureMode:       wasm.FailureModeDeny,
+		RateLimitPolicies: make([]wasm.RateLimitPolicy, 0),
+	}
+
+	t, err := r.topologyIndexesFromGateway(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	rateLimitPolicies := t.PoliciesFromGateway(gw)
+
+	logger.V(1).Info("wasmPluginConfig", "#RLPS", len(rateLimitPolicies))
+
+	// Shallow copy by assignment
+	rateLimitPoliciesSorted := rateLimitPolicies
+
+	// Sort RLPs for consistent comparison with existing objects
+	sort.Sort(kuadrantgatewayapi.PolicyByCreationTimestamp(rateLimitPoliciesSorted))
+
+	for _, policy := range rateLimitPolicies {
```

Comment:
```suggestion
	// Sort RLPs for consistent comparison with existing objects
	sort.Sort(kuadrantgatewayapi.PolicyByCreationTimestamp(rateLimitPolicies))

	for _, policy := range rateLimitPolicies {
```

Is there a need for shallow copy in order to sort since it's looped using the intial `rateLimitPolicies` anyway ? 🤔

Reply:
not really. Updated